File: distribution/python/dqops/client/models/similar_table_model.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="SimilarTableModel")


@_attrs_define
class SimilarTableModel:
    """Model that describes a table that is similar to a reference table. Similar tables are used to build the data lineage
    graph.

        Attributes:
            difference (Union[Unset, int]): Table similarity score. Lower numbers indicate higher similarity.
            connection_name (Union[Unset, str]): Connection name
            schema_name (Union[Unset, str]): Schema name
            table_name (Union[Unset, str]): Table name
    """

    difference: Union[Unset, int] = UNSET
    connection_name: Union[Unset, str] = UNSET
    schema_name: Union[Unset, str] = UNSET
    table_name: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        difference = self.difference
        connection_name = self.connection_name
        schema_name = self.schema_name
        table_name = self.table_name

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if difference is not UNSET:
            field_dict["difference"] = difference
        if connection_name is not UNSET:
            field_dict["connection_name"] = connection_name
        if schema_name is not UNSET:
            field_dict["schema_name"] = schema_name
        if table_name is not UNSET:
            field_dict["table_name"] = table_name

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        difference = d.pop("difference", UNSET)

        connection_name = d.pop("connection_name", UNSET)

        schema_name = d.pop("schema_name", UNSET)

        table_name = d.pop("table_name", UNSET)

        similar_table_model = cls(
            difference=difference,
            connection_name=connection_name,
            schema_name=schema_name,
            table_name=table_name,
        )

        similar_table_model.additional_properties = d
        return similar_table_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/core/similarity/SimilarTableModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */

package com.dqops.core.similarity;

import com.dqops.metadata.sources.PhysicalTableName;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;
import lombok.EqualsAndHashCode;

/**
 * Model that describes a table that is similar to a reference table. Similar tables are used to build the data lineage graph.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode
@ApiModel(value = "SimilarTableModel", description = "Model that describes a table that is similar to a reference table. Similar tables are used to build the data lineage graph.")
public class SimilarTableModel {
    /**
     * Table similarity score. Lower numbers indicate higher similarity.
     */
    @JsonPropertyDescription("Table similarity score. Lower numbers indicate higher similarity.")
    private int difference;

    /**
     * A similarity score as a percentage. A value 100.0 means that the tables are probably equal.
     */
    @JsonPropertyDescription("A similarity score as a percentage. A value 100.0 means that the tables are probably equal.")
    private double similarityPct;

    /**
     * Connection name
     */
    @JsonPropertyDescription("Connection name")
    private String connectionName;

    /**
     * Schema name
     */
    @JsonPropertyDescription("Schema name")
    private String schemaName;

    /**
     * Table name
     */
    @JsonPropertyDescription("Table name")
    private String tableName;

    public SimilarTableModel() {
    }

    public SimilarTableModel(int difference, String connectionName, PhysicalTableName physicalTableName) {
        this.difference = difference;
        this.similarityPct = ((DataSimilarityFormula.WORD_COUNT * 64) - difference) * 100.0 / (DataSimilarityFormula.WORD_COUNT * 64);
        this.connectionName = connectionName;
        this.schemaName = physicalTableName.getSchemaName();
        this.tableName = physicalTableName.getTableName();
    }
}
